Prevent not category links in breadcrumbs at product page

vovayatsyuk opened the pull request:
### Description
Magento 2.2.4 only.
When a product is opened from a homepage and menu has a link to the homepage,
two home links will be inserted into breadcrumbs.

**Update: Added the case with "Contacts" link in the navigation.**

### Manual testing scenarios
1. Open `app/code/Magento/Theme/view/frontend/templates/html/topmenu.phtml`
2. Insert the following code:

```php
<li class="level0 level-top" role="presentation">
    <a href="<?php echo $block->getUrl() ?>" class="level-top" tabindex="-1" role="menuitem"><span>Home</span></a>
</li>
<li class="level0 level-top" role="presentation">
    <a href="<?php echo $block->getUrl('contact') ?>" class="level-top" tabindex="-1" role="menuitem"><span>Contacts</span></a>
</li>
```

Right before 

```php
<?= /* @escapeNotVerified */ $_menu ?>
```

3. Open homepage or contacts page
4. Click on any product to navigate to its page (It may be a product in cart)
5. Breadcrumbs will have two "Home" links, or Contacts page will be added into breadcrumbs.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

vovayatsyuk committed:
Prevent double home links in breadcrumbs.

This may happen if menu has a link to the homepage, and
product is opened from the homepage

VladimirZaets commented:
Hi, @vovayatsyuk , I took your PR into processing, thank you for collaboration.

VladimirZaets commented on app/code/Magento/Catalog/view/frontend/web/js/product/breadcrumbs.js:
```diff
@@ -193,6 +193,11 @@ define([
                     }
+
+                    // prevent double home link, when menu has link to the home
+                    if (categoryUrl === BASE_URL) {
```
First of all, you should change BASE_URL to window.BASE_URL according to Magento JS code style. 
Also, we don't want use global variables, will be great if you pass base URL to the widget configuration.

vovayatsyuk committed:
Get rid of global variable usage

VladimirZaets commented:
@vovayatsyuk thanks for fast update

vovayatsyuk committed:
Improved logic to skip other non-category page links

vovayatsyuk commented:
Updated PR, as the same issue happens with other non-category links that may be added into navigation (Contacts, for example).

vovayatsyuk commented:
It still doesn’t work well, if non-category links where added programmatically into the tree.

They will have ‘nav-‘ css class just like category items and latest logic will detect them as category links.

:(

It looks like there is no way to detect if item from navigation is a category item or not.

Any thoughts?

vovayatsyuk commented:
### Propose

Revert the changes and do the following instead:

1. Add `is_category_item` flag into the [Node](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Catalog/Plugin/Block/Topmenu.php#L159-L166)
2. Add `category-item` CSS class when [rendering menu](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Theme/Block/Html/Topmenu.php#L305) if Node has `is_category_item` flag
3. Use `category-item` CSS class to detect if link found in navigation by breadcrumbs.js script is a category link.

This logic will cover both cases: links added manually into template and links added programmatically by third-party modules.

> We can skip step 1 and use [Node ID](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Catalog/Plugin/Block/Topmenu.php#L161)(`category-node-123`) in step 2 to check if this a category item. If it starts with `category-node-`, then it's a category, otherwise - custom link.

What do you think?

vovayatsyuk committed:
Use 'category-item' links only.

This change prevents adding custom links, like 'contacts', 'home', etc.into breadcrumbs

vovayatsyuk commented:
I've rewritten this PR completely because the previous solution does not work
with programmatically added items.

vovayatsyuk committed:
Fixed failed test

vovayatsyuk committed:
Fixed maximum line length validation error

VladimirZaets commented:
Hi @vovayatsyuk.
> What do you think?
- Yes, I looked the code, I think the solution is good, thanks.

magento-engcom-team commented:
Hi @vovayatsyuk. Thank you for your contribution. 
Changes from your Pull Request will be available with the upcoming 2.2.5 release.

Please, consider porting this solution across release lines.
You may use [Porting tool](https://github.com/magento/magento2/wiki/Simple-way-to-port-your-solutions-automatically) to port commits automatically.
